components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfSignUpManager.java

```diff
@@ -0,0 +1,298 @@
+        notificationResponseBean.setUser(user);
+
+        IdentityMgtMessageContext messageContext = new IdentityMgtMessageContext();
+
+        Map<String, Object> props = new HashMap<>();
+        props.put(IdentityStoreConstants.USER, user);
+        props.put(SELF_SIGN_UP_PROPERTIES, properties);
+
+        Event event = new Event(SELF_SIGN_UP_EVENT, props);
+
+        try {
+
+            eventService.pushEvent(event, messageContext);
+
+            String confirmationCode = (String) messageContext.getParameter(CONFIRMATION_CODE);
+
+            // If notification internally managed, the confirmation code will be null.
+            notificationResponseBean.setCode(confirmationCode);
+        } catch (IdentityException e) {
+            throw Utils.handleServerException(ErrorCodes.TRIGGER_NOTIFICATION,
+                                              user.getUniqueUserId(), e);
+        }
+        return notificationResponseBean;
+    }
+
+    /**
+     * Check whether user is already confirmed or not.
+     *
+     * @param uniqueUserId Unique ID of the user.
+     * @return True if the user is already confirmed.
+     * @throws IdentityRecoveryException If error occurs while user registration.
+     */
+    public boolean isUserConfirmed(String uniqueUserId) throws IdentityRecoveryException {
+
+        if (StringUtils.isBlank(uniqueUserId)) {
+
+            throw new IdentityRecoveryClientException(ErrorCodes.INVALID_USER_ID.getCode(),
+                                                      ErrorCodes.INVALID_USER_ID.getMessage());
+        }
+
+        UserRecoveryDataStore userRecoveryDataStore = JDBCRecoveryDataStore.getInstance();
+        UserRecoveryData load = userRecoveryDataStore.loadByUserUniqueId(uniqueUserId);
+
+        boolean isUserConfirmed = false;
+
+        if (load == null || !RecoveryScenarios.SELF_SIGN_UP.equals(load.getRecoveryScenario())) {
+            isUserConfirmed = true;
+        }
+        return isUserConfirmed;
+    }
+
+    /**
+     * Confirm self sign-up of a user.
+     *
+     * @param code Confirmation code to be validated.
+     * @throws IdentityRecoveryException If error occurs while user registration.
+     */
+    public void confirmUserSelfSignUp(String code) throws IdentityRecoveryException {
+
+        if (!config.isSelfSignUpEnabled()) {
+            throw Utils.handleClientException(ErrorCodes.DISABLE_SELF_SIGN_UP);
+        }
+
+        UserRecoveryDataStore userRecoveryDataStore = JDBCRecoveryDataStore.getInstance();
+
+        // If load method returns data, it means the code is validated. Otherwise it returns exceptions.
+        UserRecoveryData recoveryData = userRecoveryDataStore.loadByCode(code);
+        String uniqueId = recoveryData.getUserUniqueId();
+
+        if (!RecoverySteps.CONFIRM_SIGN_UP.equals(recoveryData.getRecoveryStep())) {
+            throw Utils.handleClientException(ErrorCodes.INVALID_CODE);
```

Comment:
Better to have context (i.e. confirmation code, user unique id) into the exception.